### PR TITLE
[RESTEASY-1987] keep independent list of Future objects to cancel on …

### PR DESCRIPTION
**Pull request by rsearls**
…shutdownNow request

### CODE REVIEW

File: resteasy-client/src/main/java/org/jboss/resteasy/plugins/providers/sse/client/SseEventSourceImpl.java [hunk 1]
```diff
@@ -291,7 +294,15 @@ private void internalClose()
             });
          }
       }
-      executor.shutdownNow();
+
+      // showdownNow only the objects registered by this class
+      for(Future<?> f : futureEventHandlerList) {
+         if (!f.isDone()) {
+            f.cancel(true);
```
Comment (jimma) [inline resteasy-client/src/main/java/org/jboss/resteasy/plugins/providers/sse/client/SseEventSourceImpl.java]: This looks good to me. Just two minor things: 
From Future javadoc, looks f.cancel() checks if the task is already completed. Can we remove f.isDone() check?
After cancel the future,  do we still need executor.awaitTermination(timeout, unit)?

Comment (rsearls) [reply resteasy-client/src/main/java/org/jboss/resteasy/plugins/providers/sse/client/SseEventSourceImpl.java]: I've removed f.isDone.
We don't want to call executor.awaitTermination(timeout, unit).  If the user is using executor for other tasks we do not want canceling the EventHandler task to terminate all other task.  The user's code should call executor.awaitTermination(timeout, unit) before exiting his app.

**Comment by rsearls** [comment]
@jimma I think I've addressed concerns with these changes.   What do you think?

File: resteasy-client/src/main/java/org/jboss/resteasy/plugins/providers/sse/client/SseEventSourceImpl.java [hunk 2]
```diff
@@ -255,10 +258,9 @@ public boolean close(final long timeout, final TimeUnit unit)
       internalClose();
       try
       {
-         if (!executor.awaitTermination(timeout, unit))
-         {
-            return false;
-         }
+         // mimic executor.awaitTermination giving time for tasks to complete
+         long duration = TimeUnit.MILLISECONDS.convert(timeout, unit);
+         Thread.sleep(duration);
```
Comment (jimma) [inline resteasy-client/src/main/java/org/jboss/resteasy/plugins/providers/sse/client/SseEventSourceImpl.java]: Is it possible to run internalClose() with schduleExecutor.scheduleWithFixedDelay() without thread.sleep?   Is it better ?

**Comment by rsearls** [comment]
@jimma I made the change requested.   I have a concern that here is no type of exception that will be thrown that allows us to catch and execute the "onErrorConsumers" block of code.   That code has been removed from this particular code change.  What are your thoughts on this?

**Comment by rsearls** [comment]
@asoldano, @jimma I've made the change from array to variable, however with both solutions I am getting random failures on org.jboss.resteasy.test.providers.sse.SseSmokeTest when running TravisCI.  In some runs 1 configuration will fail.  In other runs multiple configurations will fail.  There is no consistent failure on a JDK version or wfly version.  It appears the registered event does not get a chance to run before it is cancel.  I have not found a solution to resolve this timing issue.  What would you guys suggest?

**Comment by NicoNes** [comment]
@rsearls, @jimma, @asoldano I allowed myself to propose another solution to fix problem describe in the JIRA.
Let me know what you think about it: https://github.com/resteasy/Resteasy/pull/1677

**Comment by asoldano** [comment]
@rsearls considering the comments above, are you ok with us going with #1677 changes and closing this PR?

**Comment by rsearls** [comment]
Yes.  Jim's solution works.  Closing this PR.

**Pull request closed without merging**